### DISCUSSION TITLE
Support for ~Copyable enums and associated types

### DISCUSSION THREAD

**Discussion by Jnosh** [pointfreeco/swift-case-paths#206] [Ideas]
CasePaths doesn't currently support non-copyable types.

I don't currently have a strong enough need for this to explore it more myself but decided to open this discussion so others that hit this limitation could chime in here if needed.

Some features of CasePaths, like accessing an associated type, won't work for ~Copyable types so only a subset of the regular API could be offered for such enums and enum cases.

**Comment by mbrandonw** [comment]
Hi @Jnosh, thanks for bringing this up! It's certainly something worth exploring, but also as far as I can tell even just plain key paths don't work with non-copyable types. In fact, it crashes the compiler:

```swift
struct User: ~Copyable {
  var name = ""
}

func foo() {
  let user = User()
  let _ = user.name              // ✅
  let _ = user[keyPath: \.name]  // 💥 Compiler crash
}
```

So, looks like we've got a ways to go before we are going to have case paths working with non-copyable types.

**Reply by Jnosh** [reply]
😅 Yeah, non-copyable types probably need to stabilize a bit before this even becomes worth exploring in earnest.

There are also many (known) limitations, one I've hit occasionally is that they can't be used in tuples yet which is  also likely to cause issues with cases with multiple payloads.

Thank you for the very quick response though!
